openmls/src/tree/sender_ratchet.rs

```diff
@@ -53,152 +57,192 @@ impl Default for SenderRatchetConfiguration {
+#[cfg_attr(any(feature = "test-utils", test), derive(PartialEq))]
+pub(crate) struct RatchetSecret {
+    secret: Secret,
+    generation: Generation,
+}
+
+impl RatchetSecret {
+    /// Create an initial [`RatchetSecret`] with `generation = 0` from the given
+    /// [`Secret`].
+    pub(crate) fn initial_ratchet_secret(secret: Secret) -> Self {
         Self {
-            index,
+            secret,
             generation: 0,
-            past_secrets: vec![secret.clone()],
         }
     }
+
+    /// Return the generation of this [`RatchetSecret`].
+    pub(crate) fn generation(&self) -> Generation {
+        self.generation
+    }
+
+    /// Consume this [`RatchetSecret`] to derive a pair of [`RatchetSecrets`],
+    /// as well as the [`RatchetSecret`] of the next generation and return both.
+    pub(crate) fn ratchet_forward(
+        &mut self,
+        backend: &impl OpenMlsCryptoProvider,
+        ciphersuite: &Ciphersuite,
+    ) -> Result<(Generation, RatchetKeyMaterial), SecretTreeError> {
+        let nonce = derive_tree_secret(
+            &self.secret,
+            "nonce",
+            self.generation,
+            ciphersuite.aead_nonce_length(),
+            backend,
+        )?;
+        let key = derive_tree_secret(
+            &self.secret,
+            "key",
+            self.generation,
+            ciphersuite.aead_key_length(),
+            backend,
+        )?;
+        self.secret = derive_tree_secret(
+            &self.secret,
+            "secret",
+            self.generation,
+            ciphersuite.hash_length(),
+            backend,
+        )?;
+        let generation = self.generation;
+        self.generation += 1;
+        Ok((
+            generation,
+            (AeadKey::from_secret(key), AeadNonce::from_secret(nonce)),
+        ))
+    }
+}
+
+/// [`SenderRatchet`] used to derive key material for decryption. It keeps the
+/// [`RatchetKeyMaterial`] of epochs around until they are retrieved. This
+/// behaviour can be configured via the `out_of_order_tolerance` and
+/// `maximum_forward_distance` of the given [`SenderRatchetConfiguration`].
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[cfg_attr(any(feature = "test-utils", test), derive(PartialEq))]
+pub struct DecryptionRatchet {
+    past_secrets: VecDeque<Option<RatchetKeyMaterial>>,
+    ratchet_head: RatchetSecret,
+}
+
+impl DecryptionRatchet {
+    /// Creates e new SenderRatchet
+    pub fn new(secret: &Secret) -> Self {
+        Self {
+            past_secrets: VecDeque::new(),
+            ratchet_head: RatchetSecret::initial_ratchet_secret(secret.clone()),
+        }
+    }
+
+    /// Remove elements from the `past_secrets` queue until it is within the
+    /// bounds determined by the [`SenderRatchetConfiguration`].
+    fn prune_past_secrets(&mut self, configuration: &SenderRatchetConfiguration) {
+        while self.past_secrets.len() > configuration.out_of_order_tolerance() as usize {
```

Comment:
We decided to use something like the `MessageSecretsStore::resize` function does to avoid `while` loops. I have no strong opinions on this. `while` is more concise, but also more error prone and harder to debug.

Reply:
Ah, I was not aware that we had made a decision to avoid `while` functions generally. I'll take a look.

Reply:
Right, last time I was the one pushing for it. I still think not having `while` loops is generally preferable, because if they loop forever in production it can be nasty.